### ISSUE TITLE
Loss function goes to nan after n number of iterations - might be problem of exploding gradient?

### ISSUE THREAD

**Issue by issararab** [Marcovaldong/lstmp.pytorch#2]
Hi @Marcovaldong ,

I had to implement an LSTM with projection because of the high performance shown in the literature when dealing with quite long and complex sequences where we need a larger number of units in the cell state passed to the following cells to carry complex information through the intermediate cells.

I've used the sample provided by https://pytorch.org/blog/optimizing-cuda-rnn-with-torchscript/ to implement an optimized LSTMCell as well. It turned out to be exactly similar to your implementation.

Since you have experience with LSTMs with Projection, do you know the reason why after  a given number of iterations (let's say 1200 iterations) we start having a "nan" loss of the cross-entropy.

I have also tried your implementation and same thing happened. One issue that might cause that is the exploding gradient during training. I tried to use gradient clipping with a max norm of 5.0 but this didn't solve the problem. Do you have any idea or improvement we can add here?

Best regards,
AI

**Comment by issararab** [comment]
orm of 1.0.

**Comment by issararab** [comment]
To train on long sequences, just clip the gradients to a max norm of 1.0.
